stdlib/condition.mli

```diff
@@ -13,41 +13,112 @@
+   and is scheduled,
+   the property [P] may be falsified by some other thread
+   that is able to acquire the mutex [m] and alter the data structure [D].
+
+   Assuming that [D] is a shared data structure
+   protected by the mutex [m],
+   and assuming that [c] is a condition variable
+   associated with the mutex [m]
+   and with the property [P],
+   a typical usage scenario is as follows:
    {[
      Mutex.lock m;
-     while (* some predicate P over D is not satisfied *) do
+     while (* the property P over D is not satisfied *) do
        Condition.wait c m
      done;
      (* Modify D *)
-     if (* the predicate P over D is now satisfied *) then Condition.signal c;
+     if (* the property P over D is now satisfied *) then Condition.signal c;
      Mutex.unlock m
    ]}
+
+   {!signal} and {!broadcast} should be called only when the mutex [m]
+   associated with the condition variable [c] is locked.
+   Otherwise, there would be a risk that some other thread is currently
+   inside a critical section, has determined that the property [P]
+   associated with [c] is false, but has not yet called [wait]. In such a
+   case, the signal would be lost, leading to a potential deadlock.
 *)
 
 type t
 (** The type of condition variables. *)
 
 val create : unit -> t
-(** Return a new condition variable. *)
+(**[create()] creates and returns a new condition variable.
+   This condition variable should be associated (in the programmer's mind)
+   with a certain mutex [m], and should be used only when the mutex [m] is
+   locked. *)
 
 val wait : t -> Mutex.t -> unit
-(** [wait c m] atomically unlocks the mutex [m] and suspends the
-   calling process on the condition variable [c]. The process will
-   restart after the condition variable [c] has been signalled.
-   The mutex [m] is locked again before [wait] returns. *)
+(**The call [wait c m] is permitted only if [m] is the mutex associated
+   with the condition variable [c], and only if [m] is currently locked.
+   This call atomically unlocks the mutex [m] and suspends the
+   current thread on the condition variable [c]. This thread can
+   later be woken up after the condition variable [c] has been signaled
+   via {!signal} or {!broadcast}; however, it can also be woken up for
+   no reason. The mutex [m] is locked again before [wait] returns. One
+   cannot assume that the property [P] associated with the condition
+   variable [c] holds when [wait] returns; one must explicitly test
+   whether [P] holds after calling [wait]. *)
 
 val signal : t -> unit
-(** [signal c] restarts one of the processes waiting on the
-   condition variable [c]. *)
+(**[signal c] wakes up one of the threads waiting on the condition
+   variable [c]. This call is permitted only if the mutex [m]
+   associated with [c] is currently locked. It should be the case that
+   the property [P] associated with the condition variable [c] is true
+   when [signal] is called. *)
```

Comment:
I don't think this is right.  Since the waiting threads will all re-test the property P when woken up, the condition variable can be signaled at any time and in any state of the shared data.  Moreover, between the time thread 1 signals the condition and unlocks the mutex, and the time thread 2 is woken up from the wait and reacquires the mutex, a thread 3 may have acquired the mutex, falsified property P, and released the mutex.

So, signaling "only when the property P holds" is an optimization, to avoid too many useless wake-ups of waiting threads, but not a strong requirement.

Reply:
I have removed this requirement.